source/change-streams/tests/README.rst

```diff
@@ -158,4 +158,7 @@ The following tests have not yet been automated, but MUST still be tested
 #. The ``killCursors`` command sent during the "Resume Process" must not be allowed to throw an exception.
 #. ``$changeStream`` stage for ``ChangeStream`` against a server ``>=4.0`` that has not received any results yet MUST include a ``startAtOperationTime`` option when resuming a changestream.
 #. ``ChangeStream`` will resume after a ``killCursors`` command is issued for its child cursor.
-
+#. ``ChangeStream`` against a server ``>=4.2`` must track the latest ``postBatchResumeToken`` included in an ``aggregate`` or ``getMore`` response if available.
+#. ``ChangeStream`` against a server ``>=4.2`` must resume with the ``resumeAfter`` set to ``_id`` of the most recent document returned when the cursor is in the middle of a batch.
```

Comment:
This may also be a question for @kevinAlbs, but why would we ever resume before 

A batch consists of zero or more documents returned by the preceding `aggregate` or `getMore` command. Given that a resumable error occurs when a `getMore` operation fails, why or how would the driver ever resume in the _middle_ of a batch (i.e. while iterating locally cached documents)? Based on [Resume Process](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#resume-process), `postBatchResumeToken` always takes priority over other resume tokens or `startAtOperationTime`. [Response Format](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#response-format) also suggests that a 4.2+ response will always include `postBatchResumeToken`.

I think the intention here is that we should test that the [resume token accessor method](https://github.com/mongodb/specifications/blob/master/source/change-streams/change-streams.rst#exposing-all-resume-tokens) returns the token we expect. Consider the following:

 * First batch from `aggregate` response
  * If the batch is empty or we've iterated to the last element, the accessor should return the `postBatchResumeToken` for a 4.2+ server and fall back to `startAfter` or `resumeAfter` options for pre-4.2 servers. In the absense of all of those options, there is nothing to return (while `startAtOperationTime` is useful for resuming, it's not applicable for this method).
  * If the batch is not empty but we haven't iterated at all, I don't think the spec is clear about what we should return. Do we peak at the first document in the batch and extract its document resume token (i.e. `_id`)?
  * If the batch is not empty and we've iterated up to but _not_ including the last element, return the cached document resume token
 * Second or later batch from `getMore` response
 * If the batch is empty, we've iterated to the last element, or it's not empty and we've iterated up to but _not_ including the last element, follow the corresponding rule as above
 *  If the batch is not empty but we haven't iterated at all, I think we're supposed to return the `postBatchResumeToken` from the previous command response (falling back to fall back to `startAfter` or `resumeAfter` for pre-4.2 servers), per the first bullet under `aggregate` above.

Does this make sense? /cc: @kevinAlbs

Reply:
Agree with all of your points. We can't test resuming in the middle of iterating a batch because there's no way to cause the driver to resume then. We can only check the resume token returned by getResumeToken (or resume token event).

> Response Format also suggests that a 4.2+ response will always include postBatchResumeToken

Yes, 4.2+ change streams will always include a postBatchResumeToken. From an email sent by Bernard with subject "Change Stream High Water Mark Upgrade/Downgrade & Drivers Summary":

> Batches returned by change streams from 4.0.7+ or 4.2 will always include a postBatchResumeToken field within the cursor object, which is guaranteed to be equal to or later than the resume token of the final document in the batch.

> If the batch is not empty but we haven't iterated at all, I don't think the spec is clear about what we should return. Do we peak at the first document in the batch and extract its document resume token (i.e. _id)?

I created https://jira.mongodb.org/browse/SPEC-1263 to clarify this (though I think this can be done after tests). In that case, we should return the original `startAfter` or `resumeAfter` or an unset optional.

Reply:
Done